featurizers/map4.py

```diff
@@ -0,0 +1,128 @@
+"""
+Code inspired by the original work of the authors of the MAP4 Fingerprint:
+https://github.com/reymond-group/map4
+"""
+
+import itertools
+from collections import defaultdict
+
+import numpy as np
+from mhfp.encoder import MHFPEncoder
+from rdkit.Chem import MolToSmiles, PathToSubmol
+from rdkit.Chem.rdchem import Mol
+from rdkit.Chem.rdmolops import FindAtomEnvironmentOfRadiusN, GetDistanceMatrix
+
+from featurizers.minhash import Minhash
+
+
+def _find_env(Mol: Mol, idx: int, radius: int) -> str:
+    """
+    Function for getting the environment for given atom. By the environment, we mean the structures,
+    that are adjacent to the given atom.
+
+    :param Mol: - given molecule
+    :param idx: - index of the heavy atom in the molecule of which an environment will be found
+    :param radius: - maximum radius of search from the given atom
+    :return: string in SMILES format
+    """
+    env = FindAtomEnvironmentOfRadiusN(Mol, idx, radius)
+    atom_map = {}
+
+    submol = PathToSubmol(Mol, env, atomMap=atom_map)
+
+    if idx in atom_map:
+        smiles = MolToSmiles(
+            submol,
+            rootedAtAtom=atom_map[idx],
+            canonical=True,
+            isomericSmiles=False,
+        )
+        return smiles
+    return ""
+
+
+def _get_atom_envs(Mol: Mol, radius: int):
+    """
+    For each atom get its environment.
+
+    :param Mol:
+    :param radius:
+    :return:
+    """
+    atoms_env = {}
+    for atom in Mol.GetAtoms():
+        idx = atom.GetIdx()
+        for r in range(1, radius + 1):
+            if idx not in atoms_env:
+                atoms_env[idx] = []
+            atoms_env[idx].append(_find_env(Mol, idx, r))
+    return atoms_env
+
+
+def _all_pairs(Mol: Mol, atoms_envs: dict, radius: int, is_counted: bool):
+    """
+    Gets a list of atom-pair molecular shingles - circular structures written as SMILES, separated by the bond distance
+    between the two atoms along the shortest path.
+
+    :param Mol:
+    :param atoms_envs:
+    :param radius:
+    :param is_counted:
+    :return:
+    """
+    atom_pairs = []
+    distance_matrix = GetDistanceMatrix(Mol)
+    num_atoms = Mol.GetNumAtoms()
+    shingle_dict = defaultdict(int)
+
+    for idx_1, idx_2 in itertools.combinations(range(num_atoms), 2):
+        dist = str(int(distance_matrix[idx_1][idx_2]))
+
+        for i in range(radius):
+            env_a = atoms_envs[idx_1][i]
+            env_b = atoms_envs[idx_2][i]
+
+            ordered = sorted([env_a, env_b])
+
+            shingle = "{}|{}|{}".format(ordered[0], dist, ordered[1])
```

Comment:
Same molecules can be represented in different ways in SMILES format. Because of that, even for 2 identical molecules, their substructures may be found in a different order, because of a different atom indexing. That part forces order between substructures, which are included in shingles.
Also, changed to f-string.

Reply:
Here's a simple example with a carbon chain C2H5 - CH (CH3) - C5H11:

![sample_mol o](https://github.com/Arch4ngel21/emf/assets/92851969/6241cf6f-8cf2-4058-9968-2915aa055942)
![sample_mol_2 o](https://github.com/Arch4ngel21/emf/assets/92851969/b567e92a-6059-4972-82ce-c8c089f8bb19)